src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,576 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as RustSdkCryptoJs from "@matrix-org/matrix-sdk-crypto-wasm";
+import { OlmMachine } from "@matrix-org/matrix-sdk-crypto-wasm";
+
+import { Curve25519AuthData, KeyBackupInfo, KeyBackupSession } from "../crypto-api/keybackup";
+import { IMegolmSessionData } from "../crypto";
+import { Logger } from "../logger";
+import { ClientPrefix, IHttpOpts, MatrixError, MatrixHttpApi, Method } from "../http-api";
+import { RustBackupCryptoEventMap, RustBackupCryptoEvents, RustBackupDecryptor, RustBackupManager } from "./backup";
+import { CryptoEvent, TypedEventEmitter } from "../matrix";
+import { encodeUri, sleep } from "../utils";
+import { RustCrypto } from "./rust-crypto";
+
+/**
+ * Extract the dependence of the PerSessionKeyBackupDownloader, main reason is to make testing easier.
+ */
+export interface OnDemandBackupDelegate {
```

Comment:
I'm not convinced this "delegate" abstraction layer is helpful; rather, I think it obfuscates what's going on.

Plenty of other components (eg `BackupManager`) just take references to the `OlmMachine`, the http API, etc, and call them directly. Can't we do that here? For the tests, you need to mock out each of those dependencies, but that's not very hard?